src/robusta/utils/scope.py

```diff
@@ -0,0 +1,87 @@
+import logging
+import re
+from abc import abstractmethod, ABC
+from typing import Dict, Optional, Union, List
+
+from pydantic import BaseModel, root_validator
+
+ScopeIncludeExcludeParams = Dict[str, Optional[Union[str, List[str]]]]
+
+
+class ScopeParams(BaseModel):
+    include: Optional[List[ScopeIncludeExcludeParams]]
+    exclude: Optional[List[ScopeIncludeExcludeParams]]
+
+    @root_validator
+    def check_non_empty(cls, data: Dict) -> Dict:
+        if not (data.get("include") is not None or data.get("exclude") is not None):
+            raise ValueError("scope requires include and/or exclude subfield")
+        return data
+
+    @root_validator
+    def check_and_normalize(cls, data: Dict) -> Dict:
+        """Check and normalize entries inside include/exclude"""
+        for key in ["include", "exclude"]:
+            entry = data[key]
+            if entry is None:
+                continue
+            if entry == []:
+                raise ValueError("scope include/exclude specification requires at least one matcher")
+            for inc_exc_params in entry:
+                for attr_name, regex_or_regexes in inc_exc_params.items():
+                    if isinstance(regex_or_regexes, str):
+                        regex_or_regexes = [regex_or_regexes]
+                    inc_exc_params[attr_name] = regex_or_regexes
+        return data
+
+
+class BaseScopeMatching(ABC):
+    @abstractmethod
+    def get_scope_matching_data(self) -> Dict[str, Union[str, Dict[str, str]]]:
+        raise NotImplementedError
+
+    def scope_inc_exc_matches(self, scope_inc_exc: Optional[list]):
+        return any(self.scope_matches(scope) for scope in scope_inc_exc)
+
+    def scope_matches(self, scope: Dict[str, List[str]]):
+        # scope is e.g. {'labels': ['app=oomki.*,app!=X.*Y']}
+        # or {'name': ['pod-xyz.*'], 'title': ['fdc.*a', 'fdd.*b'], 'type': ['ISSUE']}
+        for attr_name, attr_matchers in scope.items():
+            if not self.scope_attribute_matches(attr_name, attr_matchers):
+                return False
+        return True
+
+    def scope_attribute_matches(self, attr_name: str, attr_matchers: List[str]):
+        data = self.get_scope_matching_data()
+        if attr_name not in data:
+            logging.warning(f'Scope match on non-existent attribute "{attr_name}" ({data=}')
+            return False
+        attr_value = data[attr_name]
+        for attr_matcher in attr_matchers:
+            if hasattr(self, f"scope_match_{attr_name}"):
```

Comment:
This is not clear
I would do:
```
if attr_name == "attributes":
  return self.match_path_attributes(...)
```
Then, in `match_path_attributes`, for each attribute path, run `get_attribute_by_path`, and do the matching logic here

Reply:
done